srtp/srtp.c

```diff
@@ -147,53 +147,141 @@ unsigned int srtp_get_version()
+                    template_session_keys->rtp_xtn_hdr_cipher) {
+                /* do nothing */
+            } else if (session_keys->rtp_xtn_hdr_cipher) {
+                status = srtp_cipher_dealloc(session_keys->rtp_xtn_hdr_cipher);
+                if (status)
+                    return status;
             }
 
-            if (session_keys->rtp_cipher) {
-                srtp_cipher_dealloc(session_keys->rtp_cipher);
+            /*
+             * deallocate rtcp cipher, if it is not the same as that in
+             * template
+             */
+            if (template_session_keys &&
+                session_keys->rtcp_cipher ==
+                    template_session_keys->rtcp_cipher) {
+                /* do nothing */
+            } else if (session_keys->rtcp_cipher) {
+                status = srtp_cipher_dealloc(session_keys->rtcp_cipher);
+                if (status)
+                    return status;
             }
 
-            if (session_keys->rtp_auth) {
-                srtp_auth_dealloc(session_keys->rtp_auth);
+            /*
+             * deallocate rtcp auth function, if it is not the same as that in
+             * template
+             */
+            if (template_session_keys &&
+                session_keys->rtcp_auth == template_session_keys->rtcp_auth) {
+                /* do nothing */
+            } else if (session_keys->rtcp_auth) {
+                status = srtp_auth_dealloc(session_keys->rtcp_auth);
+                if (status)
+                    return status;
             }
 
+            /*
+             * zeroize the salt value
+             */
+            octet_string_set_to_zero(session_keys->salt, SRTP_AEAD_SALT_LEN);
+            octet_string_set_to_zero(session_keys->c_salt, SRTP_AEAD_SALT_LEN);
+
             if (session_keys->mki_id) {
+                octet_string_set_to_zero(session_keys->mki_id,
+                                         session_keys->mki_size);
                 srtp_crypto_free(session_keys->mki_id);
+                session_keys->mki_id = NULL;
             }
 
-            if (session_keys->limit) {
+            /*
+             * deallocate key usage limit, if it is not the same as that in
+             * template
+             */
+            if (template_session_keys &&
+                session_keys->limit == template_session_keys->limit) {
+                /* do nothing */
+            } else if (session_keys->limit) {
                 srtp_crypto_free(session_keys->limit);
             }
         }
+    }
 
-        srtp_crypto_free(str->session_keys);
+    if (stream_template &&
+        stream->session_keys == stream_template->session_keys) {
+        /* do nothing */
+    } else {
+        srtp_crypto_free(stream->session_keys);
     }
 
-    if (str->enc_xtn_hdr) {
-        srtp_crypto_free(str->enc_xtn_hdr);
+    status = srtp_rdbx_dealloc(&stream->rtp_rdbx);
+    if (status)
+        return status;
+
+    /* DAM - need to deallocate EKT here */
```

Comment:
DAM?

Reply:
Might as well implement the EKT deallocation now?

Reply:
Talked to @pabuhler about this. EKT is not really up to scratch so work should be done to either disable it (or if someone is up to it) fix it - but that is a different chain of changes.

Reply:
And to be clear this is not a new change, this code was mostly moved from further down the file.